server/blob/blob-objectstorage/src/main/java/org/apache/james/blob/objectstorage/AESPayloadCodec.java

```diff
@@ -19,51 +19,61 @@
 import org.apache.james.blob.objectstorage.crypto.CryptoConfig;
-import org.apache.james.blob.objectstorage.crypto.PBKDF2StreamingAeadFactory;
 import org.jclouds.io.Payloads;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import com.google.crypto.tink.subtle.AesGcmHkdfStreaming;
+import com.google.crypto.tink.Aead;
+import com.google.crypto.tink.aead.AeadConfig;
+import com.google.crypto.tink.subtle.AesGcmJce;
 
 public class AESPayloadCodec implements PayloadCodec {
+
     private static final Logger LOGGER = LoggerFactory.getLogger(AESPayloadCodec.class);
-    private final AesGcmHkdfStreaming streamingAead;
+    private static final byte[] EMPTY_ASSOCIATED_DATA = new byte[0];
+    private static final int PBKDF2_ITERATIONS = 65536;
+    private static final int KEY_SIZE = 256;
+    private static final String SECRET_KEY_FACTORY_ALGORITHM = "PBKDF2WithHmacSHA256";
+
+    private final Aead aead;
 
     public AESPayloadCodec(CryptoConfig cryptoConfig) {
-        streamingAead = PBKDF2StreamingAeadFactory.newAesGcmHkdfStreaming(cryptoConfig);
+        try {
+            AeadConfig.register();
+
+            SecretKey secretKey = deriveKey(cryptoConfig);
+            aead = new AesGcmJce(secretKey.getEncoded());
+        } catch (GeneralSecurityException e) {
+            throw new RuntimeException("Error while starting AESPayloadCodec", e);
+        }
+    }
+
+    private static SecretKey deriveKey(CryptoConfig cryptoConfig) throws NoSuchAlgorithmException, InvalidKeySpecException {
+        byte[] saltBytes = cryptoConfig.salt();
+        SecretKeyFactory skf = SecretKeyFactory.getInstance(SECRET_KEY_FACTORY_ALGORITHM);
+        PBEKeySpec spec = new PBEKeySpec(cryptoConfig.password(), saltBytes, PBKDF2_ITERATIONS, KEY_SIZE);
+        return skf.generateSecret(spec);
     }
 
     @Override
-    public Payload write(InputStream is) {
-        PipedInputStream snk = new PipedInputStream();
+    public Payload write(InputStream inputStream) {
         try {
-            AtomicLong length = new AtomicLong();
-            PipedOutputStream src = new PipedOutputStream(snk);
-            OutputStream outputStream = streamingAead.newEncryptingStream(src, PBKDF2StreamingAeadFactory.EMPTY_ASSOCIATED_DATA);
-            Thread copyThread = new Thread(() -> {
-                try (OutputStream stream = outputStream) {
-                    length.addAndGet(IOUtils.copy(is, stream));
-                } catch (IOException e) {
-                    throw new RuntimeException("Stream copy failure ", e);
-                }
-            });
-            copyThread.setUncaughtExceptionHandler((Thread t, Throwable e) ->
-                LOGGER.error("Unable to encrypt payload's input stream",e)
-            );
-            copyThread.start();
-            return new Payload(Payloads.newInputStreamPayload(snk), Optional.of(length.get()));
+            byte[] ciphertext = aead.encrypt(IOUtils.toByteArray(inputStream), EMPTY_ASSOCIATED_DATA);
```

Comment:
I'm not sure we want to load entire mails into memory...

Reply:
You're right, but we have to globally rework the encryption.
I'm pretty sure that the streaming solution it is more elegant and better, but I was not able to make it work properly.

Do you agree if I open an issue for that?

Reply:
ok, it could be a good way to split this task